File: leetcode/02_add_two_numbers.ts
/* https://leetcode.com/problems/add-two-numbers/ */

/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function addTwoNumbers(l1: ListNode | null, l2: ListNode | null): ListNode | null {

    let dummyHead = new ListNode(0)
    let current = dummyHead
    let carry = 0

    // continue while we have digits to process in either number
    // or a carry from previous addition
    while (l1 || l2 || carry) {
        // get current digits (0 if no digit exists)
        const digit1 = l1?.val ?? 0  // e.g., for 342, first digit is 2
        const digit2 = l2?.val ?? 0  // e.g., for 465, first digit is 5

        // add digits and any carry from previous position
        const sum = digit1 + digit2 + carry

        // carry is 1 if sum >= 10, otherwise 0
        carry = Math.floor(sum / 10)
        
        // new digit is remainder after removing carry
        current.next = new ListNode(sum % 10)
        current = current.next

        // move to next digits
        l1 = l1?.next ?? null
        l2 = l2?.next ?? null
    }

    return dummyHead.next
};

File: leetcode/02_add_two_numbers.py
""" https://leetcode.com/problems/add-two-numbers/ """

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        # create a dummy head node
        dummy = ListNode(0)
        current = dummy
        carry = 0
        
        # iterate while we have nodes or carry
        while l1 or l2 or carry:
            # get values from nodes or use 0 if node is None
            x = l1.val if l1 else 0
            y = l2.val if l2 else 0
            
            # calculate sum and new carry
            total = x + y + carry
            carry = total // 10
            
            # create new node with ones digit
            current.next = ListNode(total % 10)
            current = current.next
            
            # move to next nodes if they exist
            l1 = l1.next if l1 else None
            l2 = l2.next if l2 else None
        
        return dummy.next
        
